multiplexer.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package pubsub
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+)
+
+// Multiplexer allows multiple subscriptions to be made sharing a single
+// message queue from the hub. This means that all the messages for the
+// various subscriptions are called back in the order that the messages were
+// published. If more than on handler is added to the Multiplexer that matches
+// any given topic, the handlers are called back one after the other in the
+// order that they were added.
+type Multiplexer interface {
+	TopicMatcher
+	Add(matcher TopicMatcher, handler interface{}) error
+}
+
+type element struct {
+	matcher  TopicMatcher
+	callback *structuredCallback
+}
+
+type multiplexer struct {
+	mu         sync.Mutex
+	outputs    []element
+	marshaller Marshaller
+}
+
+// NewMultiplexer creates a new multiplexer for the hub and subscribes it.
+// Unsubscribing the multiplexer stops calls for all handlers added.
+// Only structured hubs support multiplexer.
+func NewMultiplexer(hub Hub) (Unsubscriber, Multiplexer, error) {
+	if hub == nil {
+		return nil, nil, errors.NotValidf("nil hub")
+	}
+	// TODO: make the multiplexer work with a simple hub too.
```

Comment:
IMO you should just have two types: StructuredHubMultiplexer, and SimpleHubMultiplexer.